[ISSUE-11][시큐리티 설정] 스프링 시큐리티 설정 보완

KimHJ97 opened the pull request:
## 연관된 이슈 링크

> #11 

## PR 체크리스트

- [ ]  스프링 시큐리티 설정

## 📝작업 내용

- 필터 등록 보완
  - 필터가 스프링 빈으로 등록되는 경우 자동으로 필터 등록까지 진행됨
  - 스프링 빈이 아닌, addFilter에서 new로 필터 구현체를 주입하는 것으로 변경
- 미인증 예외, 권한 예외 핸들러 추가 (+ 관련 예외 코드 ENUM 추가)
- 내부 서버용 IP 인증 필터 추가 (1조 서버 IP)

## 💬리뷰 요구사항(선택)

> 현재 JwtAuthenticationFilter에서 토큰 정보를 통해 인증 객체를 만들도록 처리하였습니다.
> 떄문에, 실제로 UserDetailsService를 이용하지 않는데, Filter에서 UserDetailsService를 호출하여 DB 조회까지 하는게 나을까요 ?
>> 토큰 자체로 서명이 되어서 검증이 되는데, 요청시마다 SQL를 호출하는 것이 나을지 고민입니다.

KimHJ97 committed:
Fix: 스프링 시큐리티 보완

 - 필터 등록 보완
   - 필터가 스프링 빈으로 등록되는 경우 자동으로 필터 등록까지 진행됨
   - 스프링 빈이 아닌, addFilter에서 new로 필터 구현체를 주입하는 것으로 변경
 - 미인증 예외, 권한 예외 핸들러 추가 (+ 관련 예외 코드 ENUM 추가)
 - 내부 서버용 IP 인증 필터 추가 (1조 서버 IP)

Ref #11

sung3441 reviewed (approved):
AccountStatus이 추후 검증에 필요하다면 db조회도 필요할 것 같습니다!
ex) AT발급 받은 후 계정이 비활성화 됨

고생하셨습니다~

Cafemug commented on src/main/java/org/example/o2o/config/security/SecurityConfig.java:
```diff
@@ -45,7 +49,13 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
+				exceptionHandling.accessDeniedHandler(customAccessDeniedHandler)
+					.authenticationEntryPoint(customAuthenticationEntryPoint)
+			)
+			//.addFilterBefore(new IpWhitelistFilter(), UsernamePasswordAuthenticationFilter.class)
```
이 부분은 왜 주석 처리 되어있는건가요?

KimHJ97 replied:
아 지금 IP를 몰라서 IpWhitelistFilter 내부적으로 "127.0.0.1"로 상수로 넣어놔서 어느 요청이든 인증에 성공해서 일단 주석 처리해놨습니다!
그리고, IP 인증 필터로 만든 인증 객체는 실제 Account에 정보가 없이 인증 객체를 만들고 있어서 어떻게 사용할지 좀더 고민해봐야 될 것 같습니다.

Cafemug reviewed (commented):
고생하셨습니다 질문하나 남겼습니다

KimHJ97 committed:
Fix: JwtAuthenticationFilter 인증 객체 생성 로직 변경

기존에 JwtAuthenticationFilter 내부에서 토큰으로 인증 객체를 생성하였으나,
UserDetatilsService로 DB 조회까지 한 후 인증 객체를 만드는 것으로 변경
 - Security 전용 필터가 아닌 OncePerRequestFilter 구현체로 필터 내부에서 예외 발생시 DispatcherService 내부적 예외 응답으로 반환됨
 - 때문에, UserDetailsService의 loadUserByUsername 호출 부분 try catch로 ApiException 예외는 핸들링하도록 처리

Ref #11

KimHJ97 committed:
Style: 코딩 컨벤션 관련 오류 수정

KimHJ97 committed:
Fix: Account 관련 필드 추가되어 커밋

KimHJ97 committed:
Fix: Account 관련 필드 추가되어 커밋
